packages/teuchos/parameterlist/test/ParameterList/ParameterList_UnitTests.cpp

```diff
@@ -297,6 +297,28 @@ TEUCHOS_UNIT_TEST( ParameterList, set_get_int )
 }
 
 
+TEUCHOS_UNIT_TEST( ParameterList, setParametersWithModifier )
+{
+  RCP<SimpleModifier> modifier = rcp(new SimpleModifier());
+  ParameterList pl("pl");
+  //pl:
+  //  A: 1.0
+  //  SubA:
+  //    B: 2
+  pl.set("A", 1.0);
+  pl.sublist("SubA", modifier).set("B", 2);
+  ParameterList new_pl("pl");
+  new_pl.setParameters(pl);
+  TEST_ASSERT( pl == new_pl );
+  new_pl.sublist("SubA").setModifier(Teuchos::null);
+  TEST_ASSERT( pl != new_pl );
```

Comment:
To be honest, I find the `==` and `!=` operators pretty much useless for parameter lists and just rely on `haveSameValuesSorted`.  In this example, after the `setParametersNotAlreadySet` call, the two parameter lists `pl` and `new_pl` look identical when printed to stdout but internally the ordering is slightly different and the `==` call fails.  That was why I put the `haveSameValuesSorted` function in a few years ago.  Regardless, I just implemented the changes you suggested.  I also discovered that the test wasn't working the way I thought it did: when I reset the modifier (`new_pl.sublist("SubA").setModifier(Teuchos::null)`) it actually changes the modifier in both `pl` and `new_pl` since `setParameters` uses references instead of making copies of the objects.  I'm changing the test a bit now to handle this correctly.

Reply:
I just implemented some of the suggested checks.  I had to edit the ParameterListModifier class to compare modifiers.  I think that `haveSameModifiers` might need to be changed as well.

Reply:
> To be honest, I find the `==` and `!=` operators pretty much useless for parameter lists and just rely on `haveSameValuesSorted`.

I guess the problem is that it was not carefully documented, or tested, exactly what was meant by `operator==()`.  I guess that was a fail.

> When I reset the modifier (`new_pl.sublist("SubA").setModifier(Teuchos::null)`) it actually changes the modifier in both `pl` and `new_pl` since `setParameters` uses references instead of making copies of the objects.

We should look at that.  The RCPs should be unique so making one null should not touch the other RCP.

> I just implemented some of the suggested checks. I had to edit the ParameterListModifier class to compare modifiers. I think that `haveSameModifiers` might need to be changed as well.

Is this ready to review again?

Reply:
I'm still looking at `haveSameModifiers`.  It looks like it still isn't doing what I expect it to.  We probably need more ParameterListModifier tests but that might have to be a separate issue.  I'll just try to fix `haveSameModifiers` and then we can open a new issue for any other improvements.

Reply:
@bartlettroscoe I just pushed some changes to fix `haveSameModifiers` and added a small unit test.  I think this is ready for review now.